### DISCUSSION TITLE
Rust side Models?

### DISCUSSION THREAD

**Discussion by Rocketpoodle** [KDAB/cxx-qt#472] [Q&A]
Are there any examples of creating model classes in rust? It would be nice to be able to create Qt models in rust as well, having access to commonly used classes like QStringListModel, and ultimately being able to extend QAbstractItemModel and QAbstractListModel would be ideal.

It seems feasible to do this with CXX, but being able to use some of the qobject attributes to declare signals and properties (since they are QObject subclasses) would be nice. 

If there is already a way that this is done with CXX_qt, could someone point me towards an example or documentation?

**Comment by ahayzen-kdab** [comment]
Hey,

With the new 0.5 release this has been made easier :-)

Have a look at our custom base class example which uses QAbstractListModel https://github.com/KDAB/cxx-qt/blob/e46b894ffb451be81957b248c4ca90d873da12c4/examples/qml_features/rust/src/custom_base_class.rs or the documentation page in the book https://kdab.github.io/cxx-qt/book/concepts/inheritance.html

You should be able to derive from the model (eg QAbstractListModel or QStringListModel), then inherit the required methods / signals and use them in Rust.

Let us know if you have any issues !

**Reply by ahayzen-kdab** [reply]
Also note that over time we might want to create abstractions to ease creating models in Rust (eg imaging having something like a `VecModel` in the future which acts like a `Vec` in Rust but a QAbstractListModel in C++). But for now it can be done via this manual route.

**Reply by Rocketpoodle** [reply]
I appreciate the response, I missed the inheritance example with models! Thank you!
